worlds/shapez/__init__.py

```diff
@@ -0,0 +1,289 @@
+    generated patches on an infinite canvas, without the need to manage your infinite resources or to pay for building
+    your factories.
+    """
+    game = "shapez"
+    options_dataclass = ShapezOptions
+    options: ShapezOptions
+    topology_present = True
+    web = ShapezWeb()
+
+    # Placeholder values in case something goes wrong
+    location_count: int = 0
+    level_logic: List[str] = ["Cutter", "Rotator", "Painter", "Color Mixer", "Stacker"]
+    upgrade_logic: List[str] = ["Cutter", "Rotator", "Painter", "Color Mixer", "Stacker"]
+    random_logic_phase_length: List[int] = [1, 1, 1, 1, 1]
+    category_random_logic_amounts: Dict[str, int] = {"belt": 0, "miner": 1, "processors": 2, "painting": 3}
+    maxlevel: int = 25
+    finaltier: int = 8
+    included_locations: Dict[str, Tuple[str, LocationProgressType]] = {}
+    client_seed: int = 123
+    shapesanity_names: List[str] = []
+
+    base_id = 20010707
+    item_name_to_id = {name: id for id, name in enumerate(item_table.keys(), base_id)}
+    location_name_to_id = {name: id for id, name in enumerate(all_locations, base_id)}
+
+    # Universal Tracker support
+    ut_active: bool = False
+    passthrough: Dict[str, any] = {}
+    location_id_to_alias: Dict[int, str] = {}
+
+    @classmethod
+    def stage_generate_early(cls, multiworld: MultiWorld) -> None:
+        if len(shapesanity_simple) == 0:
+            init_shapesanity_pool()
+
+    def generate_early(self) -> None:
+        if hasattr(self.multiworld, "re_gen_passthrough"):
+            if "shapez" in self.multiworld.re_gen_passthrough:
+                self.ut_active = True
+                self.passthrough = self.multiworld.re_gen_passthrough["shapez"]
+                self.maxlevel = self.passthrough["maxlevel"]
+                self.finaltier = self.passthrough["finaltier"]
+                self.client_seed = self.passthrough["seed"]
+                self.level_logic = [self.passthrough[f"Level building {i+1}"] for i in range(5)]
+                self.upgrade_logic = [self.passthrough[f"Upgrade building {i+1}"] for i in range(5)]
+                self.random_logic_phase_length = [self.passthrough[f"Phase {i} length"] for i in range(5)]
+                self.category_random_logic_amounts = {cat: self.passthrough[f"{cat} category buildings amount"]
+                                                      for cat in ["belt", "miner", "processors", "painting"]}
+                return
+
+        # "MAM" goal is supposed to be longer than vanilla, but to not have more options than necessary,
+        # both goal amounts for "MAM" and "Even fasterer" are set in a single option.
+        if self.options.goal == "mam" and self.options.goal_amount < 27:
+            raise OptionError("When setting goal to 1 ('mam'), goal_amount must be at least 27 and not "
+                              + str(self.options.goal_amount.value))
```

Comment:
Would recommend including the player name in the error message, so they know which yaml to fix.